### PR TITLE
fix udp's duplicate/disordered ack don't update capacity

### PR DESCRIPTION
**Pull request by dreamedcheng**
An insert select statement will trigger UDP interconnect timeout error.

The error is happened when the packet reached receiver out of order or the acks replied to sender are all lost. Because in above cases, sender will not update connection's capacity at all. Further, it won't send new packets to sender next time due to no capacity left (in fact, receiver has received all the packets and replied acks), and hang in SendChunkUDPIFC until timeout.

To fix it, we need to update the capacity and comsumedSeq when received disordered or duplicate ack.

Refer to #8811 for more details.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@dreamedcheng Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8810)!

[Click here](https://cla.pivotal.io/sync/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8810) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@dreamedcheng Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/greenplum?repositoryId=greenplum-db/gpdb&pullRequestId=8810)!

**Comment by dreamedcheng** [comment]
See issue https://github.com/greenplum-db/gpdb/issues/8811

**Comment by bhuvnesh2703** [comment]
Could we add a description.. I see the issue #8811 but good for history purposes.

**Comment by dreamedcheng** [comment]
An insert select statement will trigger UDP interconnect timeout error.

The error is triggered when the packet reached to receiver out of order or the acks to sender are all lost. Because in above cases, sender will not update connection's capacity correctly. Further, it won't send new packets to sender next time due to no capacity left (in fact, receiver has received all the packets and replied acks), and hang in SendChunkUDPIFC until timeout.

To fix it, we need to update the capacity and comsumedSeq when received disordered or duplicate ack.

**Comment by dreamedcheng** [comment]
> Could we add a description.. I see the issue #8811 but good for history purposes.

Added

File: src/backend/cdb/motion/ic_udpifc.c [hunk 1]
```diff
@@ -4205,7 +4205,18 @@ handleAckedPacket(MotionConn *ackConn, ICBuffer *buf, uint64 now)
...
+		/* Update connection's capacity when receiving a disordered or duplicate ack
+		 * so that sender can send a packet when it calls sendBuffers next time, and will
+		 * not stick in SendChunkUDPIFC loop until timeout.
+		 */
+		if (!(buf->pkt->flags & UDPIC_FLAGS_CAPACITY))
+		{
+			ackConn->capacity += buf->pkt->seq - ackConn->consumedSeq;
+			ackConn->consumedSeq = buf->pkt->seq;
+		}
+	}
```
Comment (xiong-gang) [inline src/backend/cdb/motion/ic_udpifc.c]: It doesn't seem quite right. The acks of disorder and duplicate don't have the information on which packet has been consumed. A packet is acknowledged doesn't necessarily mean it has been consumed.
In your case, all the 4 packets have been received and acknowledged, and the sender is still sending the packet. This code snippet in function `handleDataPacket` should send a 'capacity ack' back.
```
	/* dropped ack or timeout */
	if (pkt->seq < conn->conn_info.seq)
	{
		ic_statistics.duplicatedPktNum++;
		if (DEBUG3 >= log_min_messages)
			write_log("dropped ack ? ignored data packet w/ cmd %d conn->cmd %d node %d route %d seq %d expected %d flags 0x%x",
					  pkt->icId, conn->conn_info.icId, pkt->motNodeId,
					  conn->route, pkt->seq, conn->conn_info.seq, pkt->flags);
		setAckSendParam(param, conn, UDPIC_FLAGS_ACK | UDPIC_FLAGS_CAPACITY | conn->conn_info.flags, conn->conn_info.seq - 1, conn->conn_info.extraSeq);

		return false;
	}
```

Can you reproduce the hang? Is it possible the packets don't get consumed indeed? You can debug the receiver side for that information.

Comment (dreamedcheng) [reply src/backend/cdb/motion/ic_udpifc.c]: Thanks for the reply.

I just ignored the logic you mentioned above.  handleDataPacket is used to receive packets from network, and the packets are consumed in processIncomingChunks. The error I encountered may produced by an old residual trx on segment that held a tuple lock, so new trx cannot consume the tuple packets. 

I will close this PR and the corresponding issue.

**Pull request closed without merging**